internal/core/application/service.go

```diff
@@ -1320,6 +1048,417 @@ func (s *Service) handleInternalAddressEventChannel(eventsCh <-chan client.Addre
+	invoiceAmount, gotPreimageHash, err := utils.DecodeInvoice(swap.Invoice)
+	if err != nil {
+		return "", fmt.Errorf("failed to decode invoice: %v", err)
+	}
+
+	if !bytes.Equal(preimageHash, gotPreimageHash) {
+		return "", fmt.Errorf("invalid preimage hash: expected %x, got %x", preimageHash, gotPreimageHash)
+	}
+	if invoiceAmount != amount {
+		return "", fmt.Errorf("invalid invoice amount: expected %d, got %d", amount, invoiceAmount)
+	}
+
+	vhtlcAddress, _, vhtlcOpts, err := s.getVHTLC(
+		ctx,
+		nil,
+		senderPubkey,
+		gotPreimageHash,
+		nil,
+		&common.RelativeLocktime{Type: common.LocktimeTypeBlock, Value: swap.TimeoutBlockHeights.UnilateralClaim},
+		&common.RelativeLocktime{Type: common.LocktimeTypeBlock, Value: swap.TimeoutBlockHeights.UnilateralRefund},
+		&common.RelativeLocktime{Type: common.LocktimeTypeBlock, Value: swap.TimeoutBlockHeights.UnilateralRefundWithoutReceiver},
+	)
+	if err != nil {
+		return "", fmt.Errorf("failed to verify vHTLC: %v", err)
+	}
+
+	if swap.LockupAddress != vhtlcAddress {
+		return "", fmt.Errorf("boltz is trying to scam us, vHTLCs do not match")
+	}
+
+	// Reavel the preimage to Boltz, Boltz should return a signed refund transaction
+	// that can be used to claim the VHTLC
+	refundTxResponse, err := s.boltzSvc.RevealPreimage(swap.Id, hex.EncodeToString(preimage))
+	if err != nil {
+		return "", fmt.Errorf("failed to reveal preimage: %v", err)
+	}
+
+	go func() {
+		// Wait until invoice is paid then proceed with claiming the VHTLC
+
+		// Workaround to connect ws endpoint on a different port for regtest
+		wsClient := s.boltzSvc
+		if s.boltzSvc.URL == boltzURLByNetwork[common.BitcoinRegTest.Name] {
+			wsClient = &boltz.Api{WSURL: "http://localhost:9004"}
+		}
+
+		ws := wsClient.NewWebsocket()
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		err = ws.Connect()
+		for err != nil {
+			log.WithError(err).Warn("failed to connect to boltz websocket")
+			time.Sleep(time.Second)
+			log.Debug("reconnecting...")
+			err = ws.Connect()
+			if ctx.Err() != nil {
+				log.Warnf("timeout while connecting to websocket: %v", ctx.Err())
+				return
+			}
+		}
+
+		err = ws.Subscribe([]string{swap.Id})
+		for err != nil {
+			log.WithError(err).Warn("failed to subscribe for swap events")
+			time.Sleep(time.Second)
+			log.Debug("retrying...")
+			err = ws.Subscribe([]string{swap.Id})
+		}
+
+		for update := range ws.Updates {
+			log.Infof("WS update: %+v\n", update)
+			parsedStatus := boltz.ParseEvent(update.Status)
+
+			confirmed := false
+			switch parsedStatus {
+			case boltz.InvoiceSettled:
+				confirmed = true
+			case boltz.InvoiceFailedToPay, boltz.TransactionFailed, boltz.TransactionLockupFailed:
+				log.Warnf("something went wrong: %s", update.Status)
+				s.claimVHTLC(ctx, preimage, *vhtlcOpts)
+				return
+			}
+			if confirmed {
+				log.Infof("claiming VHTLC with preimage")
+				s.claimVHTLCByRefund(ctx, refundTxResponse.Transaction)
```

Comment:
if this fails for any reason let's fallback to `claimVHTLC`:
```suggestion
				if err := s.claimVHTLCByRefund(ctx, refundTxResponse.Transaction); err != nil {
				            log.WithError(err).Debug("something went wrong, falling back to claim alone...")
				            if err := s.claimVHTLC(ctx, preimage, *vhtlcOpts); err != nil {
				                        log.Warnf("failed to claim vhtlc: %s", err)
				            }
				}
```